examples/rawapibot.py

```diff
@@ -52,7 +52,7 @@ def echo(bot: telegram.Bot) -> None:
 
         if update.message:  # your bot can receive updates without messages
             # Reply to the message
-            update.message.reply_text(update.message.text)
+            update.message.reply_text(update.message.text)  # type: ignore[arg-type]
```

Comment:
that looks kinda wrong here but I guess we can leave it in?

Reply:
Mypy just complains about `message.text` being `Optional[str]` instead of `str`. The only alternative would be to cast it to str, which might be even more irritation for newbies

Reply:
Could this be better if we have add an assert befor this line? Just to assure mypy that there is something in `Message.text`.


```suggestion
            assert update.message.text
            update.message.reply_text(update.message.text)
```

Reply:
I would be surporised if MyPy parsed that correctly. But even if it does, personally I find an assertion superfluous here, as `MessageHandler` already checked that, i.e. adding a run time relevant line only for a type checker seems strange to me.

Reply:
`mypy` do honor type guards/checks within a function, as it helps narrow down `Union`s and `Optional`s (which is just a shortcut of `Union`s) of a variable in the flow. 

`mypy` only considers the type definition passed in to the function (i.e. `Message` for `update.message`) which is a general type definition for all possible messages. which is passed from the handler. In the dispatcher-handler pattern, if we want to tell `mypy` about the optional-ness outcome of each specific handler, we need to show that in the type definition, potentially thourgh a super long generic type definition that could definitely look much more “ugly” than a simple type guard in the processor function. 

Otherwise, mypy would take the `def my_processor(update: Update, ...):` literally, as if `my_processor` could be called by any code with any valid `Update`s.

<details><summary>Example</summary>

```python
from typing import Optional, List

class Message:
    text: Optional[str]


def process_message(m: Message):
    # With a type guard
    if m.text is not None:
        print(m.text.split())
        
    # Without a type guard
    print(m.text.split())
    

def mock_dispatcher(ms: List[Message]):
    for m in ms:
        if m.text is not None:
            process_message(m)


message0 = Message()
message0.text = None

message1 = Message()
message1.text = "Hello world."

mock_dispatcher([message0, message1])

```

Gives the following error in `mypy`:

```
main.py:13: error: Item "None" of "Optional[str]" has no attribute "split"
Found 1 error in 1 file (checked 1 source file)
```

Result ran with https://mypy-play.net/?mypy=latest&python=3.9
</details>

Furthermore, looking back at this specific piece code, the `echo` function is just looping through the `get_updates()` results without any handlers, so a check on the `None`ness of `Message.text` is needed anyway.

Reply:
On a sidenote, the opening comment of this file seems to be pointing itself for a `telegram.ext` example. Should it be something else instead?

https://github.com/python-telegram-bot/python-telegram-bot/blob/80b34811ab7cfb633d703f88d2a06df6e71d3551/examples/rawapibot.py#L4-L9

Reply:
> Furthermore, looking back at this specific piece code, the echo function is just looping through the get_updates() results without any handlers, so a check on the Noneness of Message.text is needed anyway.

Good catch, you're right!

> On a sidenote, the opening comment of this file seems to be pointing itself for a telegram.ext example. Should it be something else instead?

Ups, seems the search & replace went too far, when we moved `echobot2.py` -> `echobot.py` and the former `echobot.py` -> `rawapibot.py`